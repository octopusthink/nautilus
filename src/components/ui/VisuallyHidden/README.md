A `VisuallyHidden` is an easy way to make an element available to assistive technology, but hidden otherwise.

```jsx
import { Button, VisuallyHidden } from '@octopusthink/nautilus';

<Button navigation href="https://www.theatlantic.com/science/archive/2019/07/why-did-octopuses-become-smart/593155/">
  Continue reading
  <VisuallyHidden>"For Smart Animals, Octopuses Are Very Weird"</VisuallyHidden>
</Button>
```

## Usage

Use `VisuallyHidden` to hide an element from sighted users. This element (typically text) should be used to provide extra context to non-visual consumers (eg, screen readers, search engines, bots) of your content.

### Use this component for...

- Title text to contextualise an otherwise repeated link (eg. "Read more" links in an archive listing).
- Descriptions of content or interface elements that are not used for sighted users. A common use-case for this is text on an icon button like "Settings".

### Don't use this component for...

- Content that should not be read out to users of assistive technology.
- Form element labels. Use the `label` prop on the elements, and refrain from hiding labels.
- Alternative (`alt`) text for images. Use a caption or `alt` text instead.
- Labelling icons. Use an `<Icon>` with a `label` prop.

## Appearance

Any content wrapped in this component will appear off-screen.

## Variations

You can use the `as` prop to change the component/element used to render the `VisuallyHidden` component. By default it outputs a `<span>` tag, but if you are outputting block-level elements inside the `VisuallyHidden` component (eg. a `Paragraph`), this will result in invalid HTML. Instead, set the `as` prop to `"div"`:

```jsx
import { Paragraph, VisuallyHidden } from '@octopusthink/nautilus';

<React.Fragment>
  <Paragraph>(The content below is visually hidden; check out the code to see it.)</Paragraph>
  <VisuallyHidden as="div">
    <Paragraph>
      This is a large amount of visually-hidden text used to explain
      something to non-sighted users.
    </Paragraph>
  </VisuallyHidden>
</React.Fragment>
```

## Accessibility

### Hide content sparingly.

Avoid hiding text that is useful for sighted users to "visually simplfy" a UI. Oftentimes label text or extra context is helpful to all users. Visually hidden elements should be used sparingly.

Don't use `VisuallyHidden` as an afterthought or a band-aid for a fundamentally inaccessible component. Icon buttons are a good example of this; what seems like an obvious icon to you for an action in your UI might be confusing to others. Buttons are more accessible as both icon + text, as are many elements in a UI.

Don't overload users of assistive technologies with too much visually hidden text, especially in links that will be read out by screenreaders.

### Make sure links have unique labels.

Screenreaders give users the option to read out only the links in a page, so it's important that each link on a page has a unique label so that users can find what they're looking for quickly.

Use full titles for links rather than generic text.

**Do this:** "Read more `<VisuallyHidden>`about how tasty Diet Coke is`</VisuallyHidden>`"

**Not this:** "Read more `<VisuallyHidden>`about this subject`</VisuallyHidden>`".
