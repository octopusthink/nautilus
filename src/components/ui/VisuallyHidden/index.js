import { css } from '@emotion/core';
import PropTypes from 'prop-types';
import React from 'react';

export const VisuallyHiddenStyles = css`
  border: 0;
  clip-path: rect(0 0 0 0);
  height: 1px;
  margin: -1px;
  overflow: hidden;
  padding: 0;
  position: absolute;
  white-space: nowrap;
  width: 1px;
`;

export const VisuallyHiddenRevealStyles = css`
  clip-path: none;
  height: auto;
  margin: 0;
  overflow: auto;
  position: static;
  white-space: nowrap;
  width: auto;
`;

export const VisuallyHidden = (props) => {
  const { as, children, ...otherProps } = props;

  const Component = as;

  return (
    <Component css={VisuallyHiddenStyles} {...otherProps}>
      {children}
    </Component>
  );
};

VisuallyHidden.defaultProps = {
  as: 'span',
  children: undefined,
};

VisuallyHidden.propTypes = {
  /** Component/tag used to output the VisuallyHidden component. */
  as: PropTypes.elementType,
  /** @ignore */
  children: PropTypes.node,
};

export default VisuallyHidden;
