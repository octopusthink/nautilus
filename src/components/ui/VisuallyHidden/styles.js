import { css } from '@emotion/react';

export const VisuallyHiddenStyles = css`
  border: 0;
  clip-path: rect(0 0 0 0);
  clip-path: inset(50%);
  height: 1px;
  margin: -1px;
  overflow: hidden;
  padding: 0;
  position: absolute;
  white-space: nowrap;
  width: 1px;
`;

export const VisuallyHiddenRevealStyles = css`
  clip-path: none;
  height: auto;
  margin: 0;
  overflow: auto;
  position: static;
  white-space: normal;
  width: auto;
`;
