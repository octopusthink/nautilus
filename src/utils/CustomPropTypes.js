import { Children } from 'react';

const allowedChildren = (...types) => {
  return (propValue, key, componentName) => {
    let error = null;

    Children.forEach(propValue[key], (child) => {
      if (!types.includes(child.type)) {
        error = new Error(
          `${componentName} component only accepts the following components as children: ${types
            .map((element) => element.displayName || element)
            .join(', ')}.`,
        );
      }
    });

    return error;
  };
};

export default {
  allowedChildren,
};
