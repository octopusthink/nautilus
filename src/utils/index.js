export { default as CustomPropTypes } from './CustomPropTypes';
