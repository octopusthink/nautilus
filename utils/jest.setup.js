// import axios from 'axios';
// import AxiosMockAdapter from 'axios-mock-adapter';
// import { JSDOM } from 'jsdom';
// eslint-disable-next-line import/no-extraneous-dependencies
import 'jest-axe/extend-expect';
// eslint-disable-next-line import/no-extraneous-dependencies
import '@testing-library/jest-dom/extend-expect';

global.window.matchMedia = jest.fn().mockImplementation((query) => ({
  matches: false,
  media: query,
  onchange: null,
  addListener: jest.fn(),
  removeListener: jest.fn(),
}));

jest.spyOn(global.console, 'error');
jest.spyOn(global.console, 'warn');
jest.spyOn(global.console, 'log');
jest.spyOn(global.console, 'info');
jest.spyOn(global.console, 'debug');

beforeEach(() => {
  global.console.error.mockClear();
  global.console.warn.mockClear();
  global.console.log.mockClear();
  global.console.info.mockClear();
  global.console.debug.mockClear();
});
