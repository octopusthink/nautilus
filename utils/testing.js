// We only use react-testing-library in development, so ignore this linting
// error.
// eslint-disable-next-line import/no-extraneous-dependencies
import { render } from '@testing-library/react';
// eslint-disable-next-line import/no-extraneous-dependencies
import { createMemoryHistory } from 'history';
import PropTypes from 'prop-types';
import React from 'react';
import { Router } from 'react-router-dom';

import { Nautilus } from 'components';
import { nautilus as nautilusDefaultTheme, themePropTypes } from 'themes';

// Wrap `react-testing-library`'s `render` function with our providers.
const NautilusProviders = ({ children, theme }) => {
  return <Nautilus theme={theme}>{children}</Nautilus>;
};

NautilusProviders.defaultProps = {
  children: undefined,
  theme: nautilusDefaultTheme,
};

NautilusProviders.propTypes = {
  /** @ignore */
  children: PropTypes.node,
  /** @ignore */
  theme: themePropTypes,
};

const customRender = (ui, options) => {
  return render(ui, { wrapper: NautilusProviders, ...options });
};

// this is a handy function that I would utilize for any component
// that relies on the router being in context
export const renderWithRouter = (
  ui,
  {
    route = '/',
    history = createMemoryHistory({ initialEntries: [route] }),
  } = {},
) => {
  return {
    ...customRender(<Router history={history}>{ui}</Router>),
    // adding `history` to the returned utilities to allow us
    // to reference it in our tests (just try to avoid using
    // this to test implementation details).
    history,
  };
};

export const muteConsole = ({ times, type } = {}) => {
  Array.from({ length: times }).forEach(() => {
    global.console[type].mockImplementationOnce(() => jest.fn());
  });
};

// Export jest-axe helpers.
export { axe } from 'jest-axe';

// Re-export everything else as normal.
export * from '@testing-library/react';

// override render method
export { customRender as render };
