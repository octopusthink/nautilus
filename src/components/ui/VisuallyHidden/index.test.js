import React from 'react';

import { axe, render } from 'utils/testing';

import VisuallyHidden from '.';

describe('Heading', () => {
  it('should render a <span> by default', () => {
    const { container } = render(<VisuallyHidden>Hello</VisuallyHidden>);

    expect(container.firstChild.tagName).toEqual('SPAN');
  });

  it('should output its children', () => {
    const { getByTestId } = render(
      <VisuallyHidden>
        <span data-testid="child" />
      </VisuallyHidden>,
    );

    expect(getByTestId('child')).toBeDefined();
  });

  it('should be hidden with CSS', () => {
    const { container } = render(<VisuallyHidden>Howdy</VisuallyHidden>);

    expect(container).toMatchSnapshot();
  });

  it('should accept and pass through other props', () => {
    const { getByTestId } = render(
      <VisuallyHidden className="custom-class" data-testid="hidden">
        hello
      </VisuallyHidden>,
    );

    expect(getByTestId('hidden').classList).toContain('custom-class');
  });

  describe('accessibility', () => {
    it('should pass aXe tests', async () => {
      const { container } = render(<VisuallyHidden>Hello to assistive technology!</VisuallyHidden>);

      expect(await axe(container.innerHTML)).toHaveNoViolations();
    });
  });
});
