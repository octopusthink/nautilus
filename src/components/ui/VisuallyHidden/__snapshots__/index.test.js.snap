// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Heading should be hidden with CSS 1`] = `
.emotion-0 {
  border: 0;
  -webkit-clip-path: rect(0 0 0 0);
  clip-path: rect(0 0 0 0);
  height: 1px;
  margin: -1px;
  overflow: hidden;
  padding: 0;
  position: absolute;
  white-space: nowrap;
  width: 1px;
}

<div>
  <span
    class="emotion-0"
  >
    Howdy
  </span>
</div>
`;
